proposals/2045-column-lineage-endpoint.md

```diff
@@ -0,0 +1,143 @@
+# Proposal: Column lineage endpoint proposal
+
+Author(s): @julienledem
+
+Created: 20022-08-18
+
+Dicussion: [column lineage endpoint issue #2045](https://github.com/MarquezProject/marquez/issues/2045)
+
+## Overview
+
+### Use cases
+ - Find the current upstream dependencies of a column. A column in a dataset is derived from columns in upstream datasets.
+ - See column-level lineage in the dataset level lineage when available.
+ - Retrieve point-in-time upstream lineage for a dataset or a column. What did the lineage look like yesterday compared to today?
+
+### Existing elements
+
+- OpenLineage defines a [column-level lineage facet]- (https://github.com/OpenLineage/OpenLineage/blob/ff0d87d30ed6c9fe39472788948266a6d3190585/spec/facets/ColumnLineageDatasetFacet.md). 
+- Marquez has a lineage endpoint `GET /api/v1/lineage` that returns the current lineage graph connected to a job or a dataset
+
+### New Elements
+We propose to add the following:
+- Add column lineage to the lineage endpoint
+- A new column-lineage endpoint leveraging the column lineage facet to retrieve lineage for a given column.
+- Point-in-time upstream (dataset or column level) lineage given a version of a dataset.
+
+## Proposal
+
+### add column lineage to existing endpoint
+In the GET /lineage api, add column lineage to DATASET nodes' data
+```diff
+{
+  "id": "dataset:food_delivery:public.categories",
+  "type": "DATASET",
+  "data": {
+    "type": "DATASET",
+    "id": {
+      "namespace": "food_delivery",
+      "name": "public.categories"
+    },
+    "type": "DB_TABLE",
+    ...
+    "fields": [{
+		...
+    }],
+>   columnLineage: {
+>     "a": {
+>       inputFields: [
+>          {namespace: "ns", name: "name", "field": "a"},
+>          ... other inputs
+>       ],
+>       transformationDescription: "identical",
+>       transformationType: "IDENTITY"
+>     },
+>     "b": ... other output fields
+>   }
+  },
+  "inEdges": [{
+    "origin": "job:food_delivery:etl_orders.etl_categories",
+    "destination": "dataset:food_delivery:public.categories"
+  }],
+  "outEdges": [{
+    "origin": "dataset:food_delivery:public.categories",
+    "destination": "job:food_delivery:etl_orders.etl_orders_7_days"
+  }]
+}
+```
+
+### add a column-level-lineage endpoint:
+
+```
+GET /column-lineage?nodeId=dataset:food_delivery:public.delivery_7_days&column=a
```

Comment:
I'm proposing a different endpoint fot /column-lineage because the payload would be different, containing only datasets. I was considering that the columnLineage facet was already providing edges and that the inEdges and outEdges fields of the lineage graph became unnecessary.

To me /upstream or /downstream is not an endpoint as they are more of a filter on the lineage than a different result.

Reply:
>  I was considering that the columnLineage facet was already providing edges and that the inEdges and outEdges fields of the lineage graph became unnecessary

I would then change the payload from a graph consisting of nodes (with in/out edges that aren't really relevant), to more an array of datasets objects that don't have in / out edges as much of the metadata that is relevant for lineage, wouldn't apply here.

My thinking is this: the lineage call returns a set of nodes, but doesn't specify if they all have to be datasets, or all have to be jobs. It's generic in that way. What matters are the nodeIDs and that the `origin` and `destination` point to a node in the return node set. Adding a new node type `datasetField` would fulfill the API contract. But, like you said, weather a query is upstream or downstream can be a backend implementation that can be based on the node type:

```
GET /lineage?nodeId=datasetField:{namespace}:{dataset}:{field}
```

Reply:
Ohh man, it's great discussion although it took me 10 times reading to get to know what are you talking about. 

I tried to include the initial idea of Julien mixed with the feedback of Willy. 

Some clue design decisions:
* existing `lineage` endpoint will be enriched with column lineage as-is (column lineage facet included within dataset)
* new `column-lineage` will return column lineage graph with edges between dataset fields. It will reuse existing Graph data structure with new new `dataset_fields` type. Jobs won't be included in the graph -> single job may have tons of edges to the fields

Other: 
* graph depth can be controlled by url parameter
* downstream lineage will be turned off by default and can be turned on when requested.

Reply:
Thanks for the update @pawel-big-lebowski This looks good to me. I left a minor comment bellow